edb/server/pgrust/src/connection/mod.rs

```diff
@@ -0,0 +1,202 @@
+
+use crate::auth::{self};
+mod conn;
+pub mod dsn;
+pub mod openssl;
+mod raw_conn;
+pub(crate) mod state_machine;
+mod stream;
+pub mod tokio;
+
+pub use conn::Client;
+use dsn::HostType;
+pub use raw_conn::connect_raw_ssl;
+pub use state_machine::{Authentication, ConnectionSslRequirement};
+
+macro_rules! __invalid_state {
+    ($error:literal) => {{
+        eprintln!(
+            "Invalid connection state: {}\n{}",
+            $error,
+            ::std::backtrace::Backtrace::capture()
+        );
+        #[allow(deprecated)]
+        $crate::connection::ConnectionError::__InvalidState
+    }};
+}
+pub(crate) use __invalid_state as invalid_state;
+
+#[derive(Debug, thiserror::Error)]
+pub enum ConnectionError {
+    /// Invalid state error, suggesting a logic error in code rather than a server or client failure.
+    /// Use the `invalid_state!` macro instead which will print a backtrace.
+    #[error("Invalid state")]
+    #[deprecated = "Use invalid_state!"]
+    __InvalidState,
+
+    /// Error returned by the server.
+    #[error("Server error: {code}: {message}")]
+    ServerError {
+        code: String,
+        message: String,
+        extra: HashMap<ServerErrorField, String>,
+    },
+
+    /// The server sent something we didn't expect
+    #[error("Unexpected server response: {0}")]
+    UnexpectedServerResponse(String),
+
+    /// Error related to SCRAM authentication.
+    #[error("SCRAM: {0}")]
+    Scram(#[from] auth::SCRAMError),
+
+    /// I/O error encountered during connection operations.
+    #[error("I/O error: {0}")]
+    Io(#[from] std::io::Error),
+
+    /// UTF-8 decoding error.
+    #[error("UTF8 error: {0}")]
+    Utf8Error(#[from] std::str::Utf8Error),
+
+    /// SSL-related error.
+    #[error("SSL error: {0}")]
+    SslError(#[from] SslError),
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum SslError {
+    #[error("SSL is not supported by this client transport")]
+    SslUnsupportedByClient,
+    #[error("SSL was required by the client, but not offered by server (rejected SSL)")]
+    SslRequiredByClient,
+    #[error("OpenSSL error: {0}")]
+    OpenSslError(#[from] ::openssl::ssl::Error),
+    #[error("OpenSSL error: {0}")]
+    OpenSslErrorStack(#[from] ::openssl::error::ErrorStack),
+}
+
+/// A sufficient set of required parameters to connect to a given transport.
+#[derive(Clone, Default, derive_more::Debug)]
+pub struct Credentials {
+    pub username: String,
+    #[debug(skip)]
+    pub password: String,
+    pub database: String,
+    pub server_settings: HashMap<String, String>,
+}
+
+#[derive(Clone, Debug)]
+/// The resolved target of a connection attempt.
+pub enum ResolvedTarget {
+    SocketAddr(std::net::SocketAddr),
+    #[cfg(unix)]
+    UnixSocketAddr(std::os::unix::net::SocketAddr),
+}
+
+impl ResolvedTarget {
+    /// Resolves the target addresses for a given host.
+    pub fn to_addrs_sync(host: dsn::Host) -> Result<Vec<ResolvedTarget>, std::io::Error> {
```

Comment:
Ack, good catch. I will need to move this off-thread. The plan was originally to make use of the tokio blocking pool, but I removed tokio from this PR.

Will address this.

Reply:
To fix this, I used an executor to resolve the candidates, and in the property we drop the GIL so that we can effectively run it in parallel. The uvloop resolution works in a similar way and this should be sufficiently performant (it'll only be called for connection setup) until I move more of this code to rust.

Reply:
> an executor

Yup, this would effectively flip the `PY_ALLOW_THREADS` macros from no-op into actual overheads. But

> until I move more of this code to rust.

I assume the Python thread pool will go away after this, so I think it's fine that we have the overhead for a while.